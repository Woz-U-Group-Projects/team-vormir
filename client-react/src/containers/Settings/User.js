import React from 'react';
import PropTypes from 'prop-types';
import { Redirect } from 'react-router-dom';
import { connect } from 'react-redux';
import { Popup, Button } from 'semantic-ui-react';
import { signOut } from 'features/user/state/ui/SignOut.action';
import { isSignedIn, isDemoUser } from 'features/user/state/User.selector';

class User extends React.Component {
  render() {
    if (this.props.isSignOutComplete) return <Redirect to="/" />;

    return (
      <Popup
        on="click"
        trigger={
          <Button
            content={this.signOutButtonLabel()}
            icon={this.signOutButtonIcon()}
            labelPosition="right"
          />
        }
        header="Local data will de deleted!"
        content={
          <Button
            content="Confirm"
            negative
            style={{ marginTop: '1em' }}
            floated="right"
            loading={this.props.isSignOutRunning}
            disabled={this.props.isSignOutRunning}
            onClick={this.props.signOut}
          />
        }
      />
    );
  }

  signOutButtonLabel() {
    if (this.props.isDemo) return 'Reset demo';

    return this.props.isAuthenticated ? 'Sign out' : 'Delete data';
  }

  signOutButtonIcon() {
    if (this.props.isDemo) return 'refresh';

    return this.props.isAuthenticated ? 'sign out' : 'trash';
  }
}

User.propTypes = {
  isDemo: PropTypes.bool,
  isAuthenticated: PropTypes.bool,
  isSignOutRunning: PropTypes.bool,
  isSignOutComplete: PropTypes.bool,
  signOut: PropTypes.func
};

const mapStateToProps = state => ({
  isDemo: isDemoUser(state),
  isAuthenticated: isSignedIn(state),
  isSignOutRunning: state.user.ui.signOut.signOutState === 'REQUEST',
  isSignOutComplete: state.user.ui.signOut.signOutState === 'COMPLETE'
});

export default connect(
  mapStateToProps,
  { signOut }
)(User);
