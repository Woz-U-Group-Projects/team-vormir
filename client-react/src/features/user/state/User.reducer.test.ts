import reducer, { UserActionT } from './User.reducer';
import { setDemoUser, signInSuccess } from './User.action';

const initialState = reducer(undefined, {} as any);

it('returns default state', () => {
  expect(initialState).toMatchSnapshot();
});

it('changes flag when user is authenticated', () => {
  expect(reducer({ ...initialState }, signInSuccess() as UserActionT)).toEqual({
    ...initialState,
    isSignedIn: true
  });
});

it('changes flag when user is demo', () => {
  expect(reducer({ ...initialState }, setDemoUser() as UserActionT)).toEqual({
    ...initialState,
    isDemoUser: true
  });
});
