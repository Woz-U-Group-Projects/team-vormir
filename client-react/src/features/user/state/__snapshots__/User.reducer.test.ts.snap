// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`returns default state 1`] = `
Object {
  "isDemoUser": false,
  "isSignedIn": false,
  "ui": Object {
    "signIn": Object {
      "code": "",
      "email": "",
    },
    "signOut": Object {},
  },
}
`;
